partner_duplicate_mgmt/models/res_partner_duplicate.py

```diff
@@ -0,0 +1,167 @@
+    def merge_partners(self):
+        # Verify that a partner is checked for all lines
+        for line in self.merge_line_ids:
+            if not line.partner_1_selected and not line.partner_2_selected:
+                raise UserError(_(
+                    'Please select a value to preserve for the field %s.') %
+                    (line.duplicate_field_id.name))
+
+        # Update preserved partner fields values
+        self.update_preserved_partner()
+
+        # Call the method _merge of the crm partner merge widget
+        # _merge erases all the messages of the partner left
+        partners = self.partner_1_id | self.partner_2_id
+        base_wizard = self.env['base.partner.merge.automatic.wizard']
+        base_wizard.with_context(do_not_unlink_partner=True)._merge(
+            partners.ids, self.partner_preserved_id)
+
+        # Archive the partner which is not preserved
+        partner_to_archive = partners - self.partner_preserved_id
+        partner_to_archive.write({'active': False})
+
+        # Add a message to the chatter
+        message = _('Merged into %s') % (self.partner_preserved_id.name)
+        partner_to_archive.message_post(body=message)
+
+        # Change duplicate state
+        self.write({'state': 'merged'})
+
+    def _find_partner_duplicates(self):
+        cr = self.env.cr
+
+        min_similarity = self.env['ir.config_parameter'].get_param(
+            'partner_duplicate_mgmt.partner_name_min_similarity')
+
+        cr.execute('SELECT set_limit(%s)', (min_similarity,))
+        cr.execute("""
+            SELECT p1.id, p2.id
+            FROM res_partner p1
+            JOIN res_partner p2 ON p1.indexed_name % p2.indexed_name
+            WHERE p1.id != p2.id
+            AND ((p1.parent_id IS NOT DISTINCT FROM p2.parent_id)
+              OR (p1.parent_id IS NULL AND p1.id != p2.parent_id)
+              OR (p2.parent_id IS NULL AND p2.id != p1.parent_id)
+            )
+            AND NOT EXISTS (
+                SELECT NULL
+                FROM res_partner_duplicate d
+                WHERE (d.partner_1_id = p1.id AND d.partner_2_id = p2.id)
+                OR    (d.partner_1_id = p2.id AND d.partner_2_id = p1.id)
+            )
+            """)
+
+        return cr.fetchall()
+
+    def create_duplicates(self):
+        res = self._find_partner_duplicates()
+        duplicates_sorted = [tuple(sorted(r)) for r in res]
+        duplicates = list(set(duplicates_sorted))
+
+        for dup in duplicates:
+            self.create({'partner_1_id': dup[0], 'partner_2_id': dup[1]})
+
+    @api.multi
+    def action_resolve(self):
+        self.filtered(
+            lambda x: x.state == 'to_validate').write({
+                'state': 'resolved'})
+
+    @api.multi
+    def set_to_draft(self):
+        self.write({'state': 'to_validate'})
+
+    def _get_merge_lines(self):
+        return self.env['res.partner.merge.line'].create_merge_lines(
+            self.partner_1_id, self.partner_2_id)
```

Comment:
this method is not required, you may call `self.env['res.partner.merge.line'].create_merge_lines(..)` directly from `open_partner_merge_wizard()`